### DISCUSSION TITLE
Exposición «SPECTRE y Meltdown»: preguntas y respuestas

### DISCUSSION THREAD

**Discussion by gwolf** [unamfi/sistop-2022-1#166] [Q&A]
Abro este foro como un espacio de discusión para la exposición que prepararon los alumnos Jesús Eduardo Salazar y Jorge Luis Manzanares, «SPECTRE y Meltdown». El video está disponible en:

https://www.youtube.com/watch?v=hjpiJnA5toU

Pueden consultar también el documento de texto en que los alumnos presentan este mismo tema:

https://github.com/unamfi/sistop-2022-1/blob/main/exposiciones/ManzanaresJorge-SalazarJesus/Spectre-y-Meltdown.pdf

Además de participar aquí, les pido que llenen la pequña encuesta que está disponible en:

https://www.questionpro.com/t/ARiwuZqBvO

**Comment by Inakyord** [comment]
Hola, espero estén bien. ¿Con todo y el rendimiento perdido por los parches de seguridad, sigue siendo más eficiente la ejecución especulativa que sólo usar pipelining, o ya no se nota tanto las ganancias?

**Reply by JesusESD** [reply]
¿Qué tal compañero? Igualmente espero que te encuentres muy bien. La ejecución especulativa sigue siendo utilizada en la actualidad, a pesar de las limitaciones que se le han impuesto a raíz del descubrimiento de estas vulnerabilidades. Como bien mencionas, existe un detrimento en el rendimiento de los equipos, especialmente en los más antiguos. Sin embargo, dichas pérdidas no son suficientemente grandes como para eliminar la ejecución especulativa del todo. En adición a esto, los sistemas más nuevos han modificado ciertos aspectos de este tipo de ejecución, como el destino de las ramas indirectas, para prevenir estos ataques.

Por supuesto que se siguen notando los estragos en el rendimiento, pero poco a poco estamos aprendiendo a sobrellevarlos :)

**Reply by gwolf** [reply]
Una muy pequeña corrección: Aunque la ejecución especulativa se usa desde mediados de los ochenta, el mayor sufrimiento lo presentan los procesadores más modernos (de la última década), porque la profundidad de sus pipelines (→ la cantidad de etapas de ejecución de cada instrucción dentro del CPU) es mayor. Se ha intentado limitar esta profundidad, pero el éxito, sobre todo en x86, ha sido bastante limitado.

**Comment by CarlosGMendozaB** [comment]
Hola compañeros, muy buena exposición. Cuando te refieres a Spectre, donde su funcionalidad generalmente es engañar a una aplicación o aplicaciones, ¿Esta aplicación/es que son engañadas, suelen tener alguna característica en especial, es decir en común tener características que propicien este ataque o puede ser cualquier aplicación que puede ser atacado?

**Reply by JesusESD** [reply]
Saludos compañero! Recordemos que tanto Spectre como Meltdown se caracterizan por ser vulnerabilidades originadas en el hardware. Por lo tanto, antes de que se aplicaran los parches para limitar su alcance, prácticamente cualquier proceso ejecutado por un procesador afectado podía verse atacado.

**Reply by gwolf** [reply]
De acuerdo con lo que dice Jesús. No es que se _engañe_ a una aplicación (estás hablando de una escala demasiado pequeña, el CPU no entiende de aplicaciones), sino que se le lleva a ejecutar código que hace obvio este problema. "Entrenar" al predictor de saltos suena a algo complicado, pero es muy simple: Si tienes un ciclo tan sencillo como este:

    for(i=0; i<100; i++) 
        haz_algo();

esto se compilaría aproximadamente a algo así:

      ASIGNA (i) 0
    inicia_ciclo:
      COMPARA_MENOR (i) 100 EAX
      SALTA-SI-NO-CERO EAX fin_de_ciclo
      LLAMA_SUB haz_algo
      SALTA inicia_ciclo
    fin_de_ciclo:
      (...)

El predictor de saltos después de unos pocos ciclos _aprende_ (o _es entrenado a que_) la respuesta habitual de `COMPARA_MENOR` es =0, entonces _predice_ que se continuará hacia `LLAMA_SUB`. Cuando `i=100`, el predictor de saltos va a avanzar sobre `LLAMA_SUB`, para descartar el camino andado poquito después — pero su impacto en líneas de caché peude ser observado desde fuera.
